crates/locutus-router/src/isotonic_estimator.rs

```diff
@@ -0,0 +1,317 @@
+                *peer_location,
+                Adjustment {
+                    sum: total_adjustment,
+                    count: event_count,
+                },
+            );
+        }
+
+        IsotonicEstimator {
+            global_regression,
+            peer_adjustments,
+        }
+    }
+
+    /// Adds a new event to the estimator.
+    pub fn add_event(&mut self, event: IsotonicEvent) {
+        let route_distance = event.route_distance();
+
+        let point = Point::new(route_distance, event.result);
+
+        self.global_regression.add_points(&[point]);
+
+        let adjustment = event.result - self.global_regression.interpolate(route_distance).unwrap();
+
+        self.peer_adjustments
+            .entry(event.peer)
+            .or_insert_with(Adjustment::default)
+            .add(adjustment);
+    }
+
+    pub fn estimate_retrieval_time(
+        &self,
+        peer: &PeerKeyLocation,
+        contract_location: Location,
+    ) -> Result<f64, EstimationError> {
+        // Check if there are enough data points that the model won't produce
+        // garbage output, but users of this class must implement their own checks
+        // to ensure that the model is sufficiently accurate as this is an
+        // extremely low bar.
+        if self.global_regression.len() < MIN_POINTS_FOR_REGRESSION {
+            return Err(EstimationError::InsufficientData);
+        }
+
+        let distance: f64 = contract_location.distance(&peer.location.unwrap()).into();
+
+        let global_estimate = self.global_regression.interpolate(distance).unwrap();
+
+        // Regression can sometimes produce negative estimates
+        let global_estimate = global_estimate.max(0.0);
+
+        Ok(self
+            .peer_adjustments
+            .get(peer)
+            .map_or(global_estimate, |peer_adjustment| {
+                let should_use_peer_adjustment =
+                    peer_adjustment.count >= MIN_POINTS_FOR_REGRESSION as u64;
+                global_estimate
+                    + if should_use_peer_adjustment {
+                        peer_adjustment.value()
+                    } else {
+                        0.0
+                    }
+            }))
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.global_regression.len()
+    }
+}
+
+pub(crate) enum EstimatorType {
+    /// Where the estimated value is expected to increase as distance increases
+    Positive,
+    /// Where the estimated value is expected to decrease as distance increases
+    Negative,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) enum EstimationError {
+    InsufficientData, // Error indicating that there is not enough data for estimation
+}
+
+/// A routing event is a single request to a peer for a contract, and some value indicating
+/// the result of the request, such as the time it took to retrieve the contract.
+#[derive(Debug, Clone)]
+pub(crate) struct IsotonicEvent {
+    // TODO: Make generic on route type
```

Comment:
any reason why we want to make this generic? are we gonna use this for something else?

Reply:
Yes, I think that comment was an old idea, removed.